File: src/core/dependencies.ts
import type { PluginDefinition } from '../models/plugin.js';
import { ValidationError } from '../utils/errors.js';

/**
 * Resolve plugin dependencies and return plugins in installation order
 * Uses topological sorting to ensure dependencies are installed before dependents
 * @param plugins - Array of plugin definitions
 * @returns Array of plugins sorted by dependency order
 * @throws ValidationError if circular dependencies are detected or dependencies are missing
 */
export function resolveDependencies(plugins: PluginDefinition[]): PluginDefinition[] {
  // Create a map for quick lookup
  const pluginMap = new Map<string, PluginDefinition>();
  for (const plugin of plugins) {
    pluginMap.set(plugin.name, plugin);
  }

  // Validate that all dependencies exist in the plugin list
  for (const plugin of plugins) {
    if (plugin.dependencies) {
      for (const depName of plugin.dependencies) {
        if (!pluginMap.has(depName)) {
          throw new ValidationError(`Plugin "${plugin.name}" depends on "${depName}", but "${depName}" is not defined in the configuration`);
        }
      }
    }
  }

  // Topological sort using Kahn's algorithm
  const inDegree = new Map<string, number>();
  const adjacencyList = new Map<string, string[]>();
  
  // Initialize in-degree count and adjacency list
  for (const plugin of plugins) {
    inDegree.set(plugin.name, 0);
    adjacencyList.set(plugin.name, []);
  }
  
  // Build the dependency graph
  for (const plugin of plugins) {
    if (plugin.dependencies) {
      for (const depName of plugin.dependencies) {
        // depName should be installed before plugin.name
        adjacencyList.get(depName)?.push(plugin.name);
        inDegree.set(plugin.name, (inDegree.get(plugin.name) || 0) + 1);
      }
    }
  }
  
  // Queue for processing (plugins with no dependencies)
  const queue: string[] = [];
  const result: PluginDefinition[] = [];
  
  // Find all plugins with no incoming edges (no dependencies)
  for (const [pluginName, degree] of inDegree) {
    if (degree === 0) {
      queue.push(pluginName);
    }
  }
  
  // Process the queue
  while (queue.length > 0) {
    const current = queue.shift()!;
    const currentPlugin = pluginMap.get(current)!;
    result.push(currentPlugin);
    
    // Remove this node from the graph and update in-degrees
    const neighbors = adjacencyList.get(current) || [];
    for (const neighbor of neighbors) {
      const newDegree = (inDegree.get(neighbor) || 0) - 1;
      inDegree.set(neighbor, newDegree);
      
      if (newDegree === 0) {
        queue.push(neighbor);
      }
    }
  }
  
  // Check for circular dependencies
  if (result.length !== plugins.length) {
    const remainingPlugins = plugins
      .filter(plugin => !result.some(resolved => resolved.name === plugin.name))
      .map(plugin => plugin.name);
    throw new ValidationError(`Circular dependency detected involving plugins: ${remainingPlugins.join(', ')}`);
  }
  
  return result;
}

/**
 * Validate plugin dependencies without sorting
 * @param plugins - Array of plugin definitions
 * @throws ValidationError if dependencies are invalid
 */
export function validateDependencies(plugins: PluginDefinition[]): void {
  // This will throw if there are issues
  resolveDependencies(plugins);
}

File: lua/necromancer/core/dependencies.lua
local errors = require("necromancer.utils.errors")

local M = {}

---Resolve plugin dependencies using topological sort (Kahn's algorithm)
---@param plugins table[] Array of plugin definitions with name, repo, commit, dependencies?
---@return table[] Sorted plugins in installation order
function M.resolve_dependencies(plugins)
  -- Create plugin map for quick lookup
  local plugin_map = {}
  for _, plugin in ipairs(plugins) do
    plugin_map[plugin.name] = plugin
  end

  -- Validate all dependencies exist
  for _, plugin in ipairs(plugins) do
    if plugin.dependencies then
      for _, dep_name in ipairs(plugin.dependencies) do
        if not plugin_map[dep_name] then
          error(errors.ValidationError(
            string.format('Plugin "%s" depends on "%s", but "%s" is not defined in the configuration',
              plugin.name, dep_name, dep_name)
          ))
        end
      end
    end
  end

  -- Initialize in-degree count and adjacency list
  local in_degree = {}
  local adjacency = {}

  for _, plugin in ipairs(plugins) do
    in_degree[plugin.name] = 0
    adjacency[plugin.name] = {}
  end

  -- Build dependency graph
  for _, plugin in ipairs(plugins) do
    if plugin.dependencies then
      for _, dep_name in ipairs(plugin.dependencies) do
        -- dep_name should be installed before plugin.name
        table.insert(adjacency[dep_name], plugin.name)
        in_degree[plugin.name] = in_degree[plugin.name] + 1
      end
    end
  end

  -- Queue for processing (plugins with no dependencies)
  local queue = {}
  local result = {}

  -- Find all plugins with no incoming edges
  for name, degree in pairs(in_degree) do
    if degree == 0 then
      table.insert(queue, name)
    end
  end

  -- Process queue
  while #queue > 0 do
    local current = table.remove(queue, 1)
    table.insert(result, plugin_map[current])

    -- Remove this node and update in-degrees
    for _, neighbor in ipairs(adjacency[current]) do
      in_degree[neighbor] = in_degree[neighbor] - 1
      if in_degree[neighbor] == 0 then
        table.insert(queue, neighbor)
      end
    end
  end

  -- Check for circular dependencies
  if #result ~= #plugins then
    local remaining = {}
    for _, plugin in ipairs(plugins) do
      local found = false
      for _, resolved in ipairs(result) do
        if resolved.name == plugin.name then
          found = true
          break
        end
      end
      if not found then
        table.insert(remaining, plugin.name)
      end
    end
    error(errors.ValidationError(
      "Circular dependency detected involving plugins: " .. table.concat(remaining, ", ")
    ))
  end

  return result
end

---Validate plugin dependencies without sorting
---@param plugins table[] Array of plugin definitions
---@return nil Throws ValidationError if dependencies are invalid
function M.validate_dependencies(plugins)
  -- This will throw if there are issues
  M.resolve_dependencies(plugins)
end

return M
